### PR TITLE
Change Namespace for Stored Script to Only Use Id

### PR DESCRIPTION
**Pull request by jdconrad**
Currently, stored scripts use a namespace of (lang, id) to be put, get, deleted, and executed.  This is not necessary since the lang is stored with the stored script.  A user should only have to specify an id to use a stored script.  This change makes that possible while keeping backwards compatibility with the previous namespace of (lang, id).  Anywhere the previous namespace is used will log deprecation warnings.  The goal is to backport this change to 5.2 and then remove the previous namespace altogether in 6.0 giving users the opportunity to change their scripts to the new namespace.

Why is this change so large?

This change seems huge, but it's not really.  The main files changed are ScriptMetaData and ScriptService with StoredScriptSource being added.  Secondarily, most of the Request files have changed a bit to support the new/old namespaces.  Many lines that have changed are added JavaDocs and some new tests.  Also maintaining back compat eats up some lines for XContent and the wire format.

What is the new behavior?

When a user specifies a stored script, that script will be stored under both the new namespace and old namespace.

Take for example script 'A' with lang 'L0' and data 'D0'.  If we add script 'A' to the empty set, the scripts map will be ["A" -- D0, "A#L0" -- D0].  If a script 'A' with lang 'L1' and data 'D1' is then added, the scripts map will be ["A" -- D1, "A#L1" -- D1, "A#L0" -- D0].

When a user deletes a stored script, that script will be deleted from both the new namespace (if it exists) and the old namespace.

Take for example a scripts map with {"A" -- D1, "A#L1" -- D1, "A#L0" -- D0}. If a script is removed specified by an id 'A' and lang null then the scripts map will be {"A#L0" -- D0}.  To remove the final script, the deprecated namespace must be used, so an id 'A' and lang 'L0' would need to be specified.

When a user gets/executes a stored script, if the new namespace is used then the script will be retrieved/executed using only 'id', and if the old namespace is used then the script will be retrieved/executed using 'id' and 'lang'

### CODE REVIEW

**Comment by clintongormley** [comment]
@jdconrad Have you added deprecation logging for when the old form (including lang) has been used?

**Comment by jdconrad** [comment]
@clintongormley Sorry about missing the deprecation tag!  I think I always miss one.  To answer your question I added deprecation logging everywhere user-input can come in from that uses lang.

File: modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml [hunk 1]
```diff
@@ -1,14 +1,20 @@
 ---
 "Indexed script":
```
Comment (martijnvg) [inline modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml]: not part of this change, but this should be renamed to `"Stored script"` :)

File: modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml [hunk 2]
```diff
@@ -28,6 +34,8 @@
   - is_false: script
 
   - do:
+     warnings:
+       - 'specifying lang [painless] as part of the url path is deprecated'
```
Comment (martijnvg) [inline modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml]: Also I think we should update the yaml specs to just describe the id and update the yaml tests accordingly.
So that the new behavior gets tested too.

We can add `get_script_deprecated.json`, `delete_script_deprecated.json` etc. to test that the old behaviour and that warnings get returned in the response.

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteStoredScriptAction.java [hunk 3]
```diff
@@ -35,24 +36,33 @@
...
-        if (registerDefaultHandlers) {
-            controller.registerHandler(DELETE, "/_scripts/{lang}/{id}", this);
-        }
-    }
 
-    protected String getScriptLang(RestRequest request) {
-        return request.param("lang");
+        // Note {lang} is actually {id} in the first handler.  It appears
+        // parameters as part of the path must be of the same ordering relative
+        // to name or they will not work as expected.
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteStoredScriptAction.java]: true... I also don't see a way to make this nicer.

File: core/src/main/java/org/elasticsearch/script/StoredScriptSource.java [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * {@link StoredScriptSource} represents user-defined parameters for a script
+ * saved in the {@link ClusterState}.
+ */
+public class StoredScriptSource extends AbstractDiffable<StoredScriptSource> implements Writeable, ToXContent {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/script/StoredScriptSource.java]: :+1: much better than script as bytes

File: core/src/test/java/org/elasticsearch/script/StoredScriptTests.java [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    public void testInvalidDelete() {
+        ResourceNotFoundException rnfe =
+            expectThrows(ResourceNotFoundException.class, () -> ScriptMetaData.deleteStoredScript(null, "test", "lang"));
+        assertThat(rnfe.getMessage(), equalTo("stored script [test] using lang [lang] does not exist and cannot be deleted"));
+
+        rnfe = expectThrows(ResourceNotFoundException.class, () -> ScriptMetaData.deleteStoredScript(null, "test", null));
+        assertThat(rnfe.getMessage(), equalTo("stored script [test] does not exist and cannot be deleted"));
+    }
+
+    public void testSourceParsing() throws Exception {
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/script/StoredScriptTests.java]: @colings86 is adding base test classes for writeable and xcontent serialization / de-serialization via #22281

That should get merged quickly. When that is in can you make this test class extend from `AbstractSerializingTestCase`. That will add random serialization tests in addition that the serialization  that exist already.

File: core/src/main/java/org/elasticsearch/script/ScriptMetaData.java [hunk 6]
```diff
@@ -18,168 +18,414 @@
...
                 case VALUE_STRING:
-                    scripts.put(key, new ScriptAsBytes(new BytesArray(parser.text())));
+                    if (id == null) {
+                        throw new ParsingException(parser.getTokenLocation(),
+                            "unexpected token [" + token + "], expected [<id>, <code>, {]");
+                    }
+
+                    int split = id.indexOf('#');
+
+                    if (split == -1) {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/script/ScriptMetaData.java]: I think we can move the split check to the place where we assign the `id` variable?

File: core/src/test/java/org/elasticsearch/script/StoredScriptTests.java (hunk 5 above)
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/script/StoredScriptTests.java]: and maybe do the same for `ScriptMetaDataTests`?

**Review by martijnvg — CHANGES_REQUESTED** [review]
In general looks good. A left a number of comments.

I think that we should add some bwc tests too. I think we can add some tests around stored scripts in the `qa/rolling-upgrade` module?

Also I think in a followup issue, the rest actions / specs should be cleanup so that only the new api remains. This change should only go into master.

**Comment by jdconrad** [comment]
@martijnvg Thank you so much for reviewing this!  I agree that the old API should be removed in master -- the plan was to get this in, backport it to version 5.2 (although maybe 5.3 is more realistic given the upcoming holidays), and then remove the old API in master where we can.
I'll get started working on your other comments pretty soon.

File: core/src/main/java/org/elasticsearch/script/ScriptMetaData.java [hunk 7]
```diff
@@ -18,168 +18,414 @@
...
+         * Take for example script 'A' with lang 'L0' and data 'D0'.  If we add script 'A' to the
+         * empty set, the scripts {@link Map} will be ["A" -- D0, "A#L0" -- D0].  If a script
+         * 'A' with lang 'L1' and data 'D1' is then added, the scripts {@link Map} will be
+         * ["A" -- D1, "A#L1" -- D1, "A#L0" -- D0].
+         * @param id The user-specified id to use for the look up.
+         * @param source The user-specified stored script data held in {@link StoredScriptSource}.
+         */
+        public Builder storeScript(String id, StoredScriptSource source) {
+            StoredScriptSource previous = scripts.put(id, source);
+            scripts.put(source.getLang() + "#" + id, source);
```
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptMetaData.java]: I thought about this as well and it certainly makes sense, but the tricky part is the user has no choice for stored templates.  I wanted to make sure that the user could still have a stored template with the same name as a script for this version so they have time to come up with an appropriate plan for the deprecation.  With the correct planning the old scheme will simply disappear when loading the scripts from cluster state metadata because we can make it ignore (and print a warning) that an old script cannot be loaded similar to how file scripts work now for scripts that do not compile.

Also, right now this technically isn't a breaking change which I think supports the user a little better (though, hopefully no one is dependent on having scripts of the same name with different languages).

File: core/src/main/java/org/elasticsearch/script/ScriptMetaData.java [hunk 8]
```diff
@@ -18,168 +18,414 @@
...
     }
 
     @Override
-    public String type() {
-        return TYPE;
+    public void writeTo(StreamOutput out) throws IOException {
+        // Version 5.2+ will output the contents of the scripts' Map using
+        // StoredScriptSource to stored the language, code, and options.
+        if (out.getVersion().onOrAfter(Version.V_5_2_0_UNRELEASED)) {
+            out.write(scripts.size());
```
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptMetaData.java]: Nice catch!  :)

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteStoredScriptAction.java (hunk 3 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteStoredScriptAction.java]: :(, but it will be removed in master anyway

File: core/src/main/java/org/elasticsearch/script/StoredScriptSource.java (hunk 4 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/StoredScriptSource.java]: :)

File: modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml (hunk 1 above)
Comment (jdconrad) [inline modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml]: Done.

File: docs/reference/migration/migrate_6_0/scripting.asciidoc [hunk 9]
```diff
@@ -1,6 +1,8 @@
 [[breaking_60_scripting_changes]]
 === Scripting changes
 
+
+
```
Comment (jdconrad) [inline docs/reference/migration/migrate_6_0/scripting.asciidoc]: Accidental.  Good catch!

File: modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml (hunk 2 above)
Comment (jdconrad) [inline modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml]: Updated the tests to the new namespace.

File: core/src/main/java/org/elasticsearch/script/ScriptMetaData.java (hunk 6 above)
Comment (jdconrad) [inline core/src/main/java/org/elasticsearch/script/ScriptMetaData.java]: split is used in both the if and else blocks

**Comment by jdconrad** [comment]
Added tests for rolling upgrade.  These will need to be updated when the code is backported to 5.3 and when 6.0 deletes the code for the old API.

File: core/src/test/java/org/elasticsearch/script/StoredScriptTests.java (hunk 5 above)
Comment (jdconrad) [inline core/src/test/java/org/elasticsearch/script/StoredScriptTests.java]: Done.

**Comment by jdconrad** [comment]
@martijnvg I think this is ready for another round of review.  I believe I addressed all the comments, and added several tests that you requested.  Thanks!

**Comment by jdconrad** [comment]
@martijnvg I added the patch you sent me to address the bwc bugs you found with one minor change --  I decided to clean up the Script readFrom and writeTo a little bit to separate out each of the read and write versions into their own if/else block as I think that will help clear up potential future confusion.  And thanks again so much for catching some of the bwc bugs!  Ready for another round of review when you have the chance.

**Review by martijnvg — APPROVED** [review]
Thx Jack! LGTM

Maybe also open an issue to track to removal of the `lang` parameter from the script/template apis and internal classes for 6.0?

**Comment by jdconrad** [comment]
@martijnvg Thanks for all the reviews here!  Added this as a follow up issue (#22887)  I will commit this code once I get master integrated properly and all the tests passing.

**Comment by clintongormley** [comment]
@jdconrad in 6.0 i assume the namespace support will no longer be deprecated but will be removed?  If so, please add a note to the migration docs for 6.0

**Comment by jdconrad** [comment]
@clintongormley Yes, it will be removed, but that requires another change to actually remove it from 6.0 that I'll try to follow up with this week.  Right now, it's still technically deprecated there, but not removed.  Once, I make the change I'll make sure to add that as a (breaking?) note to the migration docs.

**Comment by Mpdreamz** [comment]
@jdconrad should the spec for script api's in `master` be changed from:

```json
      "path": "/_scripts/{lang}",
      "paths": [ "/_scripts/{lang}", "/_scripts/{lang}/{id}" ],
```

to

```json
      "path": "/_scripts/{id}",
      "paths": [ "/_scripts/{id}" ],
```

Or has the removal not made it in yet to `master` ?
